File: paleo_workbench/workflow/provenance_graph.py
"""Unified product lifecycle provenance graph (V8 M10).

One pure-data API composing the EXISTING authorities — catalog lineage
(:mod:`paleo_workbench.catalog.lineage_graph`), run freshness
(:mod:`paleo_workbench.workflow.freshness`), constraint lifecycle pins
(:mod:`paleo_workbench.workflow.constraint_versions`) and map-product
records (:mod:`paleo_workbench.workflow.map_product`) — into the single
dependency narrative Direction A's Inspector / lineage explorer renders:

.. code-block:: text

    RAW assets → derived well/factor inputs → constraints versions
      → factor runs → factor outputs → compilation input set
      → fusion run → integrated draft → QA → map product → freeze/publish

No second authority: every node carries the id of the system that owns it
(version ids, run ids, task ids, record ids); every edge is derived from
recorded lineage, never inferred. Missing pieces are reported as ``missing``
nodes with a reason — a gap in provenance is data, not an exception.
"""
from __future__ import annotations

from typing import Any

__all__ = ["build_product_lifecycle_graph"]


def _node(node_id: str, kind: str, label: str, **extra: Any) -> dict[str, Any]:
    node = {"id": node_id, "kind": kind, "label": label}
    node.update({k: v for k, v in extra.items() if v is not None})
    return node


def build_product_lifecycle_graph(
    project: Any,
    catalog_service: Any = None,
    *,
    product_id: str | None = None,
) -> dict[str, Any]:
    """Explain one product (or every product) from RAW to publish.

    Returns ``{"nodes": [...], "edges": [...], "gaps": [...]}`` — pure data,
    no rendering concerns. Every product contributes: its factor tasks'
    grid versions (+ freshness verdict), the constraint state each task
    pinned, the fusion versions when the product's inputs reference them,
    and the product record's own lifecycle flags (frozen/superseded/stale).
    """
    nodes: list[dict[str, Any]] = []
    edges: list[dict[str, str]] = []
    gaps: list[dict[str, str]] = []
    seen_nodes: set[str] = set()

    def add_node(node: dict[str, Any]) -> str:
        if node["id"] not in seen_nodes:
            seen_nodes.add(node["id"])
            nodes.append(node)
        return node["id"]

    def add_edge(src: str, dst: str, relation: str) -> None:
        edges.append({"source": src, "target": dst, "relation": relation})

    records = list(getattr(project, "map_products", None) or [])
    if product_id is not None:
        records = [r for r in records if str(r.id) == str(product_id)]
        if not records:
            return {
                "nodes": [],
                "edges": [],
                "gaps": [
                    {
                        "scope": f"product:{product_id}",
                        "reason": "no map product record with this id",
                    }
                ],
            }

    for record in records:
        rid = str(getattr(record, "id", "") or "")
        product_node = add_node(
            _node(
                f"product:{rid}",
                "map_product",
                str(getattr(record, "product_name", "") or rid),
                frozen=bool(getattr(record, "frozen", False)),
                status=str(getattr(record, "status", "") or ""),
                superseded_by=getattr(record, "superseded_by", None),
            )
        )

        for task in getattr(project, "factor_map_tasks", None) or []:
            if str(task.id) not in (getattr(record, "factor_task_ids", None) or []):
                continue
            grid_version = getattr(task, "grid_artifact_version_id", None)
            task_node_id = add_node(
                _node(
                    f"factor_task:{task.id}",
                    "factor_task",
                    str(task.name or task.id),
                    method=str(getattr(task, "method", "") or ""),
                    source_kind=str(getattr(task, "source_kind", "") or ""),
                    grid_version_id=str(grid_version) if grid_version else None,
                )
            )
            add_edge(product_node, task_node_id, "assembles_factor")
            if not grid_version:
                gaps.append(
                    {
                        "scope": f"factor_task:{task.id}",
                        "reason": "task has no persisted grid version "
                        "(never interpolated or artifact lost)",
                    }
                )
            else:
                version_node = add_node(
                    _node(
                        f"version:{grid_version}",
                        "factor_output",
                        f"grid {str(grid_version)[:12]}…",
                        version_id=str(grid_version),
                    )
                )
                add_edge(task_node_id, version_node, "produced")

            # constraint pins (V8 M2): the constraint state this task used
            params = dict(getattr(task, "parameters", None) or {})
            for pin in params.get("constraint_pins") or []:
                pin_group = str(pin.get("group_id") or "")
                group_node_id = add_node(
                    _node(
                        f"constraints:{pin_group}",
                        "constraints",
                        str(pin.get("group_name") or pin_group),
                        content_hash=str(pin.get("content_hash") or ""),
                        pinned_version_id=pin.get("version_id"),
                    )
                )
                add_edge(task_node_id, group_node_id, "consumed_constraints")

    # Fusion lineage: project-level section (V8 review R2 — hoisted out of
    # the per-product loop: it was duplicated per record and O(products x
    # runs)). Lazy-safe list_runs (document.runs is EMPTY pre-warm).
    if catalog_service is not None:
        try:
            runs = catalog_service.list_runs()
        except Exception:  # noqa: BLE001 — catalog access is optional
            runs = None
            gaps.append(
                {
                    "scope": "catalog",
                    "reason": "catalog runs not readable — fusion lineage omitted",
                }
            )
        if runs is not None:
            for run in runs:
                if str(run.operation) not in (
                    "factor_fusion",
                    "factor_fusion:confidence",
                    "factor_fusion:variance",
                ):
                    continue
                for version_id in run.output_version_ids or []:
                    node_id = add_node(
                        _node(
                            f"version:{version_id}",
                            "fusion_output",
                            f"{run.operation} {str(version_id)[:12]}…",
                            version_id=str(version_id),
                            run_id=str(run.id),
                        )
                    )
                    for parent in run.input_version_ids or []:
                        parent_node = add_node(
                            _node(
                                f"version:{parent}",
                                "factor_output",
                                f"input {str(parent)[:12]}…",
                                version_id=str(parent),
                            )
                        )
                        add_edge(parent_node, node_id, "fusion_input")

    return {"nodes": nodes, "edges": edges, "gaps": gaps}

File: libs/workflow_runtime/src/provenance_graph.cpp
// provenance_graph.cpp — C++ port of
// paleo_workbench/workflow/provenance_graph.py (CONV-26). See
// include/pwb/workflow_runtime/provenance_graph.hpp.

#include <pwb/workflow_runtime/provenance_graph.hpp>

#include "python_compat.hpp"

#include <set>
#include <unordered_set>

namespace pwb::workflow_runtime {

namespace {

// _node — {"id", "kind", "label"} + non-null extras.
Json make_node(const std::string& node_id, const std::string& kind,
               const std::string& label, Json extra = Json::object()) {
    Json node = Json::object();
    node["id"] = node_id;
    node["kind"] = kind;
    node["label"] = label;
    if (extra.is_object()) {
        for (auto it = extra.begin(); it != extra.end(); ++it) {
            if (it.value().is_null()) continue;
            node[it.key()] = it.value();
        }
    }
    return node;
}

}  // namespace

Json build_product_lifecycle_graph(
    const Json& project, const CatalogRepository* repository,
    const std::optional<std::string>& product_id) {
    Json nodes = Json::array();
    Json edges = Json::array();
    Json gaps = Json::array();
    std::unordered_set<std::string> seen_nodes;

    auto add_node = [&](Json node) -> std::string {
        const std::string id = node.at("id").get_ref<const std::string&>();
        if (seen_nodes.insert(id).second) {
            nodes.push_back(std::move(node));
        }
        return id;
    };
    auto add_edge = [&](const std::string& src, const std::string& dst,
                        const std::string& relation) {
        Json edge = Json::object();
        edge["source"] = src;
        edge["target"] = dst;
        edge["relation"] = relation;
        edges.push_back(std::move(edge));
    };
    auto add_gap = [&](const std::string& scope, const std::string& reason) {
        Json gap = Json::object();
        gap["scope"] = scope;
        gap["reason"] = reason;
        gaps.push_back(std::move(gap));
    };

    std::vector<Json> records;
    for (const auto& record :
         pycompat::dict_get(project, "map_products", Json::array())) {
        if (!product_id) {
            records.push_back(record);
        } else if (pycompat::str_scalar(
                       pycompat::dict_get(record, "id", Json(""))) ==
                   *product_id) {
            records.push_back(record);
        }
    }
    if (product_id && records.empty()) {
        add_gap("product:" + *product_id,
                "no map product record with this id");
        Json out = Json::object();
        out["nodes"] = std::move(nodes);
        out["edges"] = std::move(edges);
        out["gaps"] = std::move(gaps);
        return out;
    }

    for (const Json& record : records) {
        const std::string rid =
            pycompat::str_scalar(pycompat::dict_get(record, "id", Json("")));
        Json extra = Json::object();
        extra["frozen"] = pycompat::truthy(
            pycompat::dict_get(record, "frozen", Json(false)));
        extra["status"] = pycompat::str_scalar(
            pycompat::dict_get(record, "status", Json("")));
        extra["superseded_by"] =
            pycompat::dict_get(record, "superseded_by", Json(nullptr));
        const std::string product_node = add_node(make_node(
            "product:" + rid, "map_product",
            pycompat::str_scalar(pycompat::dict_get(record, "product_name",
                                                    Json("")))
                    .empty()
                ? rid
                : pycompat::str_scalar(pycompat::dict_get(
                      record, "product_name", Json(""))),
            std::move(extra)));

        std::set<std::string> factor_task_ids;
        for (const auto& tid :
             pycompat::dict_get(record, "factor_task_ids", Json::array())) {
            factor_task_ids.insert(pycompat::str_scalar(tid));
        }
        for (const auto& task :
             pycompat::dict_get(project, "factor_map_tasks", Json::array())) {
            const std::string task_id =
                pycompat::str_scalar(pycompat::dict_get(task, "id", Json("")));
            if (factor_task_ids.count(task_id) == 0) continue;
            const Json grid_version_json = pycompat::dict_get(
                task, "grid_artifact_version_id", Json(nullptr));
            const bool has_grid = !grid_version_json.is_null();
            const std::string grid_version =
                has_grid ? pycompat::str_scalar(grid_version_json) : "";

            Json task_extra = Json::object();
            task_extra["method"] = pycompat::str_scalar(
                pycompat::dict_get(task, "method", Json("")));
            task_extra["source_kind"] = pycompat::str_scalar(
                pycompat::dict_get(task, "source_kind", Json("")));
            task_extra["grid_version_id"] = has_grid
                                                ? Json(grid_version)
                                                : Json(nullptr);
            const std::string task_node_id = add_node(make_node(
                "factor_task:" + task_id, "factor_task",
                pycompat::str_scalar(pycompat::dict_get(task, "name",
                                                        Json("")))
                        .empty()
                    ? task_id
                    : pycompat::str_scalar(
                          pycompat::dict_get(task, "name", Json(""))),
                std::move(task_extra)));
            add_edge(product_node, task_node_id, "assembles_factor");

            if (!has_grid) {
                add_gap("factor_task:" + task_id,
                        "task has no persisted grid version "
                        "(never interpolated or artifact lost)");
            } else {
                Json version_extra = Json::object();
                version_extra["version_id"] = grid_version;
                const std::string version_node = add_node(make_node(
                    "version:" + grid_version, "factor_output",
                    "grid " + pycompat::head12(grid_version) + "…",
                    std::move(version_extra)));
                add_edge(task_node_id, version_node, "produced");
            }

            // Constraint pins: the constraint state this task used.
            const Json params =
                pycompat::dict_get(task, "parameters", Json::object());
            const Json pins =
                pycompat::dict_get(params, "constraint_pins", Json::array());
            if (pins.is_array()) {
                for (const auto& pin : pins) {
                    const std::string pin_group = pycompat::str_scalar(
                        pycompat::dict_get(pin, "group_id", Json("")));
                    Json pin_extra = Json::object();
                    pin_extra["content_hash"] = pycompat::str_scalar(
                        pycompat::dict_get(pin, "content_hash", Json("")));
                    pin_extra["pinned_version_id"] =
                        pycompat::dict_get(pin, "version_id", Json(nullptr));
                    const std::string group_node = add_node(make_node(
                        "constraints:" + pin_group, "constraints",
                        pycompat::str_scalar(pycompat::dict_get(
                            pin, "group_name", Json("")))
                                .empty()
                            ? pin_group
                            : pycompat::str_scalar(pycompat::dict_get(
                                  pin, "group_name", Json(""))),
                        std::move(pin_extra)));
                    add_edge(task_node_id, group_node, "consumed_constraints");
                }
            }
        }
    }

    // Fusion lineage from the catalog (repository == nullptr: Python
    // catalog_service=None — section skipped without a gap).
    if (repository != nullptr) {
        bool runs_read_failed = false;
        std::vector<RunRecord> runs;
        try {
            runs = repository->list_runs();
        } catch (...) {
            runs_read_failed = true;
            add_gap("catalog",
                    "catalog runs not readable — fusion lineage omitted");
        }
        if (!runs_read_failed) {
            for (const RunRecord& run : runs) {
                if (run.operation != "factor_fusion" &&
                    run.operation != "factor_fusion:confidence" &&
                    run.operation != "factor_fusion:variance") {
                    continue;
                }
                for (const std::string& version_id : run.output_version_ids) {
                    Json fusion_extra = Json::object();
                    fusion_extra["version_id"] = version_id;
                    fusion_extra["run_id"] = run.run_id;
                    const std::string node_id = add_node(make_node(
                        "version:" + version_id, "fusion_output",
                        run.operation + " " + pycompat::head12(version_id) +
                            "…",
                        std::move(fusion_extra)));
                    for (const std::string& parent : run.input_version_ids) {
                        Json parent_extra = Json::object();
                        parent_extra["version_id"] = parent;
                        const std::string parent_node = add_node(make_node(
                            "version:" + parent, "factor_output",
                            "input " + pycompat::head12(parent) + "…",
                            std::move(parent_extra)));
                        add_edge(parent_node, node_id, "fusion_input");
                    }
                }
            }
        }
    }

    Json out = Json::object();
    out["nodes"] = std::move(nodes);
    out["edges"] = std::move(edges);
    out["gaps"] = std::move(gaps);
    return out;
}

}  // namespace pwb::workflow_runtime
